File: frontend/lib/community.ts
import { ApiError } from "./browser-api";

export const communityRoot = "/api/auth/community";
export const publicCommunityRoot = "/api/public/community";
export const communityTabs = [
  "feed",
  "bookmarks",
  "likes",
  "notifications",
  "following",
  "blocks",
  "reports",
] as const;
export type CommunityTab = (typeof communityTabs)[number];
export type CommunityProfile = { accountId: string; displayName: string };
export type CommunityPage<T> = { items: T[]; nextBefore: number | null };
export type CommunityComment = {
  id: string;
  position: number;
  parentId: string | null;
  author: CommunityProfile | null;
  body: string;
  createdAt: string;
  deleted: boolean;
  replies: number;
  mayDelete: boolean;
};
export type SpaceParticipation = {
  following: boolean;
  accountId: string | null;
  mayParticipate: boolean;
};
export type ArticleThread = SpaceParticipation & {
  likes: number;
  liked: boolean;
  bookmarked: boolean;
  mayModerate: boolean;
  comments: CommunityPage<CommunityComment>;
};
export type CommunityArticle = {
  space: string;
  spaceName: string;
  articleId: string | null;
  route: string;
  title: string;
  authorName: string;
  createdAt: string;
};
export type SavedArticle = {
  position: number;
  article: CommunityArticle | null;
};
export type FollowedSpace = {
  position: number;
  space: string | null;
  displayName: string | null;
  available: boolean;
};
/** A comment, or a correction event with commentId null; older servers omit the correction fields. */
export type CommunityNotification = {
  position: number;
  commentId: string | null;
  actor: CommunityProfile | null;
  excerpt: string;
  article: CommunityArticle;
  createdAt: string;
  read: boolean;
  correctionId?: string | null;
  event?: "PROPOSED" | "ACCEPTED" | "DECLINED" | "STALE" | null;
};
/** What a notification says happened, after the actor's name. */
export function noticeAction(item: CommunityNotification) {
  switch (item.event) {
    case "PROPOSED":
      return `对《${item.article.title}》提了修改建议`;
    case "ACCEPTED":
      return `采纳了你对《${item.article.title}》的修改建议`;
    case "DECLINED":
      return `没有采纳你对《${item.article.title}》的修改建议`;
    case "STALE":
      return `处理时发现《${item.article.title}》已经改过，你的建议已过期`;
    default:
      return `回复了《${item.article.title}》`;
  }
}
export type CommunityReport = {
  position: number;
  commentId: string;
  reporter: CommunityProfile | null;
  author: CommunityProfile | null;
  reason: string;
  commentBody: string;
  createdAt: string;
  status: string;
};
export type BlockedAccount = {
  position: number;
  account: CommunityProfile | null;
};
export function communityMessage(error: unknown): string {
  if (!(error instanceof ApiError)) return "操作未能完成，请稍后重试。";
  const messages: Record<string, string> = {
    COMMUNITY_PARTICIPATION_REQUIRED:
      "社区成员及以上可新增互动。你仍可管理已有记录。",
    COMMUNITY_LIMIT_REACHED:
      "已达到操作频率或保存数量限制，请稍后重试，或先整理已有记录。",
    COMMUNITY_REQUEST_CONFLICT:
      "这次提交的内容与原请求不同，请刷新后核对评论。草稿已保留。",
    COMMUNITY_REPLY_UNAVAILABLE:
      "这条评论目前无法回复，请刷新后再试。草稿已保留。",
    COMMUNITY_UNAVAILABLE: "这篇文章或评论目前不可用。",
  };
  return messages[error.code ?? ""] ?? error.message;
}

File: src/main/java/io/github/core607/poketto/community/Community.java
package io.github.core607.poketto.community;

import io.github.core607.poketto.auth.AuthPrincipal;
import io.github.core607.poketto.auth.CommunityAccounts.Profile;
import java.time.Instant;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

/** Browser-account interactions resolved against current publication, never a stored article copy. */
public interface Community {
    SpaceState space(AuthPrincipal actor, String space);

    record SpaceState(boolean following, UUID accountId, boolean mayParticipate) {}

    Thread article(AuthPrincipal actor, String space, UUID articleId, UUID parentId, long before);

    void relate(AuthPrincipal actor, String space, UUID articleId, Relation relation, boolean enabled);

    void follow(AuthPrincipal actor, String space, boolean enabled);

    UUID comment(AuthPrincipal actor, String space, UUID articleId, CommentInput input);

    void deleteComment(AuthPrincipal actor, UUID commentId);

    void moderateComment(AuthPrincipal actor, UUID commentId);

    void report(AuthPrincipal actor, UUID commentId, ReportInput input);

    void block(AuthPrincipal actor, UUID accountId, boolean enabled);

    Page<SavedArticle> bookmarks(AuthPrincipal actor, long before);

    Page<SavedArticle> likes(AuthPrincipal actor, long before);

    void removeLike(AuthPrincipal actor, long position);

    void removeBookmark(AuthPrincipal actor, long position);

    void unfollow(AuthPrincipal actor, long position);

    Feed feed(AuthPrincipal actor, String cursor);

    List<FollowedSpace> following(AuthPrincipal actor);

    Page<Notification> notifications(AuthPrincipal actor, long before);

    void markRead(AuthPrincipal actor, List<Long> positions);

    Page<BlockedAccount> blocks(AuthPrincipal actor, long before);

    Page<Report> reports(AuthPrincipal actor, long before);

    void resolveReport(AuthPrincipal actor, long position, boolean remove);

    enum Relation {
        LIKE,
        BOOKMARK
    }

    record CommentInput(UUID requestId, UUID parentId, String body) {
        public CommentInput {
            Objects.requireNonNull(requestId, "comment request id is required");
            body = text(body, 4000, "comment");
        }
    }

    record ReportInput(String reason) {
        public ReportInput {
            reason = text(reason, 1000, "report reason");
        }
    }

    private static String text(String value, int maximum, String subject) {
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException(subject + " must not be blank");
        }
        if (value.codePointCount(0, value.length()) > maximum || value.indexOf('\0') >= 0) {
            throw new IllegalArgumentException(subject + " exceeds its text bounds");
        }
        for (int i = 0; i < value.length(); i++) {
            char unit = value.charAt(i);
            if (Character.isHighSurrogate(unit)) {
                if (++i >= value.length() || !Character.isLowSurrogate(value.charAt(i))) {
                    throw new IllegalArgumentException(subject + " contains invalid Unicode");
                }
            } else if (Character.isLowSurrogate(unit)) {
                throw new IllegalArgumentException(subject + " contains invalid Unicode");
            }
        }
        return value.strip();
    }

    record Page<T>(List<T> items, Long nextBefore) {
        public Page {
            items = List.copyOf(items);
        }
    }

    record Thread(
            long likes,
            boolean liked,
            boolean bookmarked,
            boolean following,
            UUID accountId,
            boolean mayParticipate,
            boolean mayModerate,
            Page<Comment> comments) {}

    record Comment(
            UUID id,
            long position,
            UUID parentId,
            Profile author,
            String body,
            Instant createdAt,
            boolean deleted,
            long replies,
            boolean mayDelete) {}

    record ArticleCard(
            String space,
            String spaceName,
            UUID articleId,
            String route,
            String title,
            String authorName,
            Instant createdAt) {}

    record SavedArticle(long position, ArticleCard article) {}

    record FollowedSpace(long position, String space, String displayName, boolean available) {}

    record Feed(List<ArticleCard> items, String nextCursor) {
        public Feed {
            items = List.copyOf(items);
        }
    }

    /**
     * Either a comment, or a correction event: {@code PROPOSED} for owners, {@code ACCEPTED},
     * {@code DECLINED} or {@code STALE} for the proposer. The other subject's fields are null.
     */
    record Notification(
            long position,
            UUID commentId,
            Profile actor,
            String excerpt,
            ArticleCard article,
            Instant createdAt,
            boolean read,
            UUID correctionId,
            String event) {}

    record BlockedAccount(long position, Profile account) {}

    record Report(
            long position,
            UUID commentId,
            Profile reporter,
            Profile author,
            String reason,
            String commentBody,
            Instant createdAt,
            String status) {}
}
